### ISSUE TITLE
SHA-1 hashing broken

### ISSUE THREAD

**Issue by wheybags** [ruslo/hunter#684]
https://shattered.io/

Should hunter maybe migrate to using a better scheme for verifying archives?
It seems cmake supports some better hashing algorithms, but the ideal would be some sort of cryptographic signing like gpg signatures.

**Comment by ruslo** [comment]
https://github.com/ruslo/hunter/issues/373#issuecomment-204984381

**Comment by wheybags** [comment]
Yeah, I agree that adding gpg signatures is a bad idea unless cmake adds
that functionality, but for now, could at least change hashes to sha256,
which is already supported by cmake.
Relying on HTTPS for security isn't really great, as the files could be
modified if your GitHub account was compromised, and still be delivered
over a perfectly valid HTTPS connection.

I'd be willing to make the changes myself, but I don't want to start
labouring on it without knowing the patch would eventually get accepted.

**Comment by ruslo** [comment]
> Relying on HTTPS for security isn't really great, as the files could be modified if your GitHub account was compromised, and still be delivered over a perfectly valid HTTPS connection

If GitHub account will be compromised then SHA256 hashes can be changed too. Can you please describe attack in details and how usage of SHA256 instead of SHA1 will help?

**Comment by wheybags** [comment]
If your GitHub account is compromised, then the hashes can be changed, sure, and that will affect anyone that upgrades their hunter version, or a new user.
However, for an existing project that already references an existing hunter version, they would be protected, because hunter gate would only accept the _exact_ same version of hunter they used before, and then hunter would only accept the _exact_ same version of the packages as it used before.
With sha1, someone could add malicious code to a hunter release, and then brute force the zip to have the same hash, so old users that would otherwise have been safe no longer are, but with a better hashing scheme, that's not feasible.

You're right, it's not perfect, as it still doesn't protect new users and anyone who upgrades, but it's still a lot better.

**Comment by ruslo** [comment]
> someone could add malicious code to a hunter release, and then brute force the zip to have the same hash

I have no control over how `*.tar.gz` release created, GitHub do it automatically. By the way SHA1 check of archive is in HunterGate module, it's not a part of this repository and need to be improved separately: https://github.com/hunter-packages/gate

One of the doable solutions will be to allow several hash checks. So "standard" SHA1 can be used for functionality and extra SHA256 to improve security. This require to support this feature in CMake ExternalProject first. Also it's still doesn't solve the problem completely since binary cache mechanism relies solely on SHA1. You can use [HUNTER_USE_CACHE_SERVERS=NO](https://docs.hunter.sh/en/latest/reference/user-variables.html#hunter-use-cache-servers) though.

**Comment by wheybags** [comment]
Sure, I imagined it would be changed in huntergate too.
And theoretically, github could be compromised too, allowing someone to modify the archives.
It's all pretty unlikely, but when the fix is so easy I don't see why it shouldn't be done. And having decent security properties will be a blocker for people getting hunter into their codebase at work in some places.
I don't see why you would want to use SHA1 and SHA256 together though? Why not just use SHA256?
Also, could the binary cache mechanism not be updated too?

**Comment by ruslo** [comment]
> And theoretically, github could be compromised too, allowing someone to modify the archives.
> If your GitHub account is compromised

If you really worried about it you should use your own server and rely on HTTPS as a security.

> when the fix is so easy I don't see why it shouldn't be done.

It's not easy. From the very beginning this project use SHA1 everywhere where it's possible. I don't think it's a trivial change.

> I don't see why you would want to use SHA1 and SHA256 together though? Why not just use SHA256?

Because there are tons of releases in `hunter.cmake` already. You want to download them all and recalculate hashes? SHA1 is for functionality, just like it used in Git itself for funtionality. As far as I aware there is no plans to remove SHA1 from Git.

> Also, could the binary cache mechanism not be updated too?

Not automagically. Somebody have to deal with it.

> And having decent security properties will be a blocker for people getting hunter into their codebase at work in some places.

I'm ready to listen more scenarios. The one that you show already doesn't have the problems with SHA1 weakness, it assumes GitHub account or service itself compromised. This can be solved by using your GitHub account or your own service. SHA1 to SHA256 is not a really solution since "tomorrow" somebody will post the "SHA256 attack example" and everything will start over again.

**Comment by wheybags** [comment]
https://plus.google.com/+LinusTorvalds/posts/7tp2gYWQugL

SHA1 is being removed from git.

Also, SHA1 has been know to be weak for years, SHA256 should be fine for a
good while.
I'm not proposing to sed s/1/256, it would be good to refactor the code to
allow people to use any hashing mechanism that cmake supports. If SHA256
 is compromised, it should be easy then to just switch to something else.

And yes, I'm not saying making this switch would be easy, but I'm
volunteering to try it myself.

**Comment by ruslo** [comment]
> https://plus.google.com/+LinusTorvalds/posts/7tp2gYWQugL
> SHA1 is being removed from git

Heh :) Okay, good to know. Removed from arguments. Other arguments still valid though.

**Comment by wheybags** [comment]
Well, yeah, it's not like it's a huge disaster to use SHA1 for now, but I would like to see the project migrate to allowing any hashing algorithm cmake supports.
So I guess the question is, if I go ahead and implement it, would you accept a patch? (not sure when I'll have time though)
I'd start with huntergate, I think.

**Comment by ruslo** [comment]
Okay to summarize the options we have:

* Modify CMake to support multiple hash-checks. SHA1 will be used for functionality and SHA256 for extra security checks. Those checks will verify only sources, hence user have to use  [HUNTER_USE_CACHE_SERVERS=NO](https://docs.hunter.sh/en/latest/reference/user-variables.html#hunter-use-cache-servers). I will accept patches as soon as they are ready.
* Move everything from SHA1 to SHA256. You can test it on your our fork. The big question is existent hashes of packages. You can move them step-by-step, I can leave the link to you repository somewhere in https://docs.hunter.sh so the users who want to use SHA256 can join your effort. I will keep it in mind but I can't promise I will apply it fast/soon.

> I'd start with huntergate, I think

Just like I said HunterGate is not tied to the Hunter so we can have SHA256 there without doing any modification to Hunter. However it need to be checked carefully.

**Comment by wheybags** [comment]
>SHA1 will be used for functionality and SHA256 for extra security checks.

Why not just use SHA256? Using both together seems crazy (except during some sort of transition period)

>Just like I said HunterGate is not tied to the Hunter so we can have SHA256 there without doing any modification to Hunter. However it need to be checked carefully.

Ok, I'll give that a go some time soon then.

**Comment by ruslo** [comment]
> Why not just use SHA256? Using both together seems crazy (except during some sort of transition period)

Because it doesn't break old code and is pretty orthogonal to current implementation. Also it's easy extendable - you can add any hash checks that implemented in CMake. For the full SHA1 to SHA256 move see [Move everything from SHA1 to SHA256](https://github.com/ruslo/hunter/issues/684#issuecomment-282559730) part.
